### PR TITLE
feat(ci.jenkins.io) add a new Kubernetes Cloud for `bom` builds

### PR DESCRIPTION
**Pull request by dduportal**
Related to https://github.com/jenkins-infra/helpdesk/issues/3521

This PR adds a nbew Kubernetes Cloud named `cik8s-bom` to allow running `bom` builds in a separated set of resources.

Why a new Kuberenetes Cloud? 

Because it's a first iteration and it allows:
- Clear separation of concern with logical separation
- Enforced pod limit on Jenkins side (Kubernetes pod limit side defined in https://github.com/jenkins-infra/kubernetes-management/pull/3846)

This PR contains 3 atomic changes (1 commit per change to ease code review):

- A fix on the template of `jenkins::controller` profile to allow defining a Kubernetes Cloud without predefined templates (required to add the new `cik8s-bom` Kubernetes Cloud)
- A new feature  (with backward compatibility implemented through hieradata settings) on the template of `jenkins::controller` profile to allow defining the default namespace of each Kubernetes Cloud through hieradata setup
- The hierdata change on ci.jenkins.io t oadd the new Kubernetes Cloud (depends on the 2 changes above)

Tested locally on a vagrant VM (+ added manually temporarly credentials): the pipeline below

```
podTemplate(
  cloud: 'cik8s-bom',
  yaml: '''
apiVersion: "v1"
kind: "Pod"
spec:
  containers:
    - name: jnlp
      image: jenkinsciinfra/jenkins-agent-ubuntu-22.04@sha256:35420f09777717099416fb80f334f5c5d34d09ac29743c12be21dbb0e1628f1f
      imagePullPolicy: "IfNotPresent"
      command:
        - "/usr/local/bin/jenkins-agent"
      env:
        - name: "PATH"
          value: "/opt/jdk-11/bin:/home/jenkins/.asdf/shims:/home/jenkins/.asdf/bin:/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin:/usr/games:/usr/local/games:/snap/bin"
        - name: "ARTIFACT_CACHING_PROXY_PROVIDER"
          value: "aws"
        - name: "JENKINS_AGENT_WORKDIR"
          value: "/home/jenkins/agent"
        - name: "JENKINS_JAVA_OPTS"
          value: "-XX:+PrintCommandLineFlags"
        - name: "JENKINS_JAVA_BIN"
          value: "/opt/jdk-11/bin/java"
      resources:
        limits:
          memory: "8Gi"
          cpu: "4"
        requests:
          memory: "8Gi"
          cpu: "4"
      securityContext:
        privileged: false
      tty: false
      volumeMounts:
        - mountPath: "/home/jenkins/.m2/repository"
          name: "volume-1"
          readOnly: false
        - mountPath: "/tmp"
          name: "volume-0"
          readOnly: false
  restartPolicy: "Never"
  volumes:
  - emptyDir:
      medium: "Memory"
    name: "volume-0"
  - emptyDir:
      medium: ""
    name: "volume-1"
  nodeSelector:
    ci.jenkins.io/agents-density: 23
  tolerations:
    - key: "ci.jenkins.io/bom"
      operator: "Equal"
      value: "true"
      effect: "NoSchedule"
'''
) {
  node(POD_LABEL) {
    stage('test') {
      withEnv(['JAVA_HOME=/opt/jdk-17']) {
        sh 'mvn -v'
        sh 'free -m'
        sh 'grep -c processor /proc/cpuinfo'
      }
    }
  }
}
```

results in:

```
+ mvn -v
Apache Maven 3.9.1 (2e178502fcdbffc201671fb2537d0cb4b4cc58f8)
Maven home: /usr/share/apache-maven-3.9.1
Java version: 17.0.6, vendor: Eclipse Adoptium, runtime: /opt/jdk-17
Default locale: en_US, platform encoding: UTF-8
OS name: "linux", version: "5.10.173-154.642.amzn2.x86_64", arch: "amd64", family: "unix"
[Pipeline] sh
+ free -m
               total        used        free      shared  buff/cache   available
Mem:          382649        1407      369859           1       11382      378997
Swap:              0           0           0
[Pipeline] sh
+ grep -c processor /proc/cpuinfo
96
```

### CODE REVIEW

File: hieradata/clients/azure.ci.jenkins.io.yaml [hunk 1]
```diff
@@ -174,8 +184,8 @@ profile::jenkinscontroller::jcasc:
...
-
-        # max_capacity: 150 # Max 50 workers (16 CPU / 32 G) with 3 pods (3*4 CPU / 3*8G) each
+        defaultNamespace: jenkins-agents
+        # max_capacity: 50 # Max 16 workers (16 CPU / 32 G) with 3 pods (4 CPU / 8G) each
         max_capacity: 50
```
Comment (lemeurherve) [inline hieradata/clients/azure.ci.jenkins.io.yaml]: To be coherent with the comment numbers:
```suggestion
        max_capacity: 48 # Max 16 workers (16 CPU / 32 G) with 3 pods (4 CPU / 8G) each
```

File: hieradata/clients/azure.ci.jenkins.io.yaml [hunk 2]
```diff
@@ -103,13 +103,23 @@ profile::jenkinscontroller::jcasc:
...
-
+        defaultNamespace: jenkins-agents
         # keep as comment below, if needed this week
         ## TODO: Temp. capacity reduction as per https://github.com/jenkins-infra/helpdesk/issues/3421#issuecomment-1477559811
         # max_capacity: 30 # Max 10 workers (16 CPU / 32 G) with 3 pods (3*4 CPU / 3*8G) each
```
Comment (lemeurherve) [inline hieradata/clients/azure.ci.jenkins.io.yaml]:
```suggestion
```

**Review by lemeurherve — DISMISSED** [review]

**Comment by dduportal** [comment]
- Puppet agent disabled on ci.jenkins.io (progressive deployment)
- Credential `cik8s-jenkins-agent-bom-sa-token` created on `ci.jenkins.io` (as https://github.com/jenkins-infra/kubernetes-management/pull/3847 was merged and applied). The token was retrieved with the following command: ` kubectl -n jenkins-agents-bom get secret jenkins-agent-token -o yaml | yq -r '.data.token' | base64 --decode`

- Comment updated, following @lemeurherve 's review

Ready to roll!

**Review by lemeurherve — APPROVED** [review]
🚀

**Comment by dduportal** [comment]
Note to self: as reminded by @lemeurherve , we'll have to update the updatecli manifest in charge of keeping pod quotas (kuberenetes sides) from jenkins quotas (ref. https://github.com/jenkins-infra/kubernetes-management/blob/main/updatecli/updatecli.d/charts/kubernetes-pods-quotas.yaml) once merged (and validated)

**Comment by dduportal** [comment]
Puppet diff looks good:

```diff
Info: Using environment 'production'
Info: Retrieving pluginfacts
Info: Retrieving plugin
Info: Retrieving locales
Info: Loading facts


Info: Applying configuration version '1681743582'
Notice: /Stage[main]/Profile::Jenkinscontroller/File[/var/lib/jenkins/casc.d/clouds.yaml]/content: 
--- /var/lib/jenkins/casc.d/clouds.yaml 2023-04-13 17:01:36.472207067 +0000
+++ /tmp/puppet-file20230417-24919-11q5g54      2023-04-17 15:00:04.946665659 +0000
@@ -392,6 +392,32 @@
             - key: "ARTIFACT_CACHING_PROXY_PROVIDER"
               value: "aws"
   - kubernetes:
+      name: "cik8s-bom"
+      containerCap: 345
+      credentialsId: "cik8s-jenkins-agent-bom-sa-token"
+      directConnection: true
+      namespace: "jenkins-agents-bom"
+      serverUrl: "<redacted>"
+      serverCertificate: |
+          <redacted>
+  - kubernetes:
       name: "cik8s"
       containerCap: 150
       credentialsId: "cik8s-jenkins-agent-sa-token"
@@ -696,7 +722,7 @@
               mountPath: "/home/jenkins/.m2/repository"
   - kubernetes:
       name: "doks"
-      containerCap: 50
+      containerCap: 48
       credentialsId: "doks-jenkins-agent-sa-token"
       directConnection: true
       namespace: "jenkins-agents"

Notice: /Stage[main]/Profile::Jenkinscontroller/File[/var/lib/jenkins/casc.d/clouds.yaml]/content: current_value '{md5}37de650d001a92aa7c424ebe016141ec', should be '{md5}7b0cc865dc9a00c6ffc23db50e8873a3' (noop)
Info: /Stage[main]/Profile::Jenkinscontroller/File[/var/lib/jenkins/casc.d/clouds.yaml]: Scheduling refresh of Exec[perform-jcasc-reload]
Notice: /Stage[main]/Profile::Jenkinscontroller/Exec[perform-jcasc-reload]: Would have triggered 'refresh' from 1 event
Notice: Class[Profile::Jenkinscontroller]: Would have triggered 'refresh' from 2 events
Notice: Stage[main]: Would have triggered 'refresh' from 1 event
Notice: Applied catalog in 15.28 seconds
```

**Pull request merged**